### PR TITLE
Adding Subcheck module in ODE

### PR DESCRIPTION
**Pull request by Mohitbalwani26**
#### References to other Issues or PRs

SEE #18403 

#### Release Notes

NO ENTRY

### CODE REVIEW

**Comment by Mohitbalwani26** [comment]
@oscarbenjamin for Now i have just added `sub_func_doit`. when i was trying to add `checkodesol` few errors popped up so i am still figuring out that as other modules also use `checkodesol` from ode.py what additional modifications do i need to make in setup.py i couldn't get that?

**Comment by oscarbenjamin** [comment]
What other modules use checkodesol?

**Comment by Mohitbalwani26** [comment]
@oscarbenjamin my bad, when i tried to move `checkodesol` it showed error in` init.py` i thought it was from diophantine.py but later on i corrected that mistake.

Are there any further changes which i should commit on this PR?

File: sympy/solvers/ode/__init__.py [hunk 1]
```diff
@@ -1,6 +1,6 @@
-from .ode import (allhints, checkinfsol, checkodesol, classify_ode,
+from .ode import (allhints, checkinfsol, classify_ode,
         constantsimp, dsolve, homogeneous_order, infinitesimals)
-
+from .subscheck import checkodesol
 __all__ = [
```
Comment (oscarbenjamin) [inline sympy/solvers/ode/__init__.py]: There was a blank line here. Can you add it back?

File: sympy/solvers/ode/ode.py [hunk 2]
```diff
@@ -346,35 +347,6 @@
...
-    ...- 1/(x**2*(z + 1/x)**2)
-    """
-    reps= {func: new}
-    for d in eq.atoms(Derivative):
-        if d.expr == func:
-            reps[d] = new.diff(*d.variable_count)
-        else:
-            reps[d] = d.xreplace({func: new}).doit(deep=False)
-    return eq.xreplace(reps)
-
```
Comment (oscarbenjamin) [inline sympy/solvers/ode/ode.py]: How many blank lines are left? Does it match the spacing between functions in other parts of the module?

File: sympy/solvers/ode/subscheck.py [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+    """
+    reps= {func: new}
+    for d in eq.atoms(Derivative):
+        if d.expr == func:
+            reps[d] = new.diff(*d.variable_count)
+        else:
+            reps[d] = d.xreplace({func: new}).doit(deep=False)
+    return eq.xreplace(reps)
+
+def checkodesol(ode, sol, func=None, order='auto', solve_for_func=True):
```
Comment (oscarbenjamin) [inline sympy/solvers/ode/subscheck.py]: I think there should be two blank lines between these functions

**Comment by oscarbenjamin** [comment]
Can you create a new file `test_subcheck.py` in ode/tests and add the tests for these functions there?

**Comment by Mohitbalwani26** [comment]
> Can you create a new file `test_subcheck.py` in ode/tests and add the tests for these functions there?

okay i am trying for that and will add blank lines wherever you suggested.

**Comment by oscarbenjamin** [comment]
Sorry I should have been clearer. When I said "add the tests for these functions there" what I meant was this:
1. Copy the tests for these functions that are already in test_ode.py over into test_subscheck.py
2. Delete the same tests from test_ode.py since we don't need to have two copies.

**Comment by Mohitbalwani26** [comment]
@oscarbenjamin i couldn't understand why the travis is failing. can you please look at this?

**Comment by Mohitbalwani26** [comment]
@oscarbenjamin i have squashed all the commits including the changes you suggested.

**Comment by oscarbenjamin** [comment]
I think this looks good.

Can you add the blank line back? https://github.com/sympy/sympy/pull/18540#discussion_r373866558

**Comment by Mohitbalwani26** [comment]
@oscarbenjamin i have added the blank line.

**Comment by oscarbenjamin** [comment]
Looks good. Merging

**Pull request merged**